docs/dynamic-component-loading.md

```diff
@@ -0,0 +1,48 @@
+
+In lot of these cases the component which is to be loaded dynamically has a non-trivial amount of dependencies which are unknown to the app itself. So the loading mechanism has to be able to resolve them.  
+The SDK tooling uses the `.deps.json` format to describe dependencies of a component, so it would be beneficial if the dynamic load was able to use this same format for dependency resolution.
+
+## Dynamic loading with dependencies
+We propose to add a new public API which would dynamically load a component with these properties:
+* Component is loaded in isolation from the app (and other components) so that potential collisions are not an issue
+* Component can use `.deps.json` to describe its dependencies. This includes the ability to describe additional NuGet packages, RID-specific and/or native dependencies.
+* Component can chose to rely on the app for certain dependencies by not including them in its `.deps.json`
+* Optionally such component can be enabled for unloading
+
+Public API (early thinking):
+* `static Assembly Assembly.LoadFileWithDependencies` - in its core similar to `Assembly.LoadFile` but it supports resolving dependencies through `.deps.json` and such. Just like `Assembly.LoadFile` it provides isolation, but also for the dependencies.
+* `static AssemblyLoadContext AssemblyLoadContext.CreateForAssemblyWithDependencies(string assemblyPath)` - "advanced" version which would return an ALC instance and not just the main assembly. Could have overloads for enabling unloadability.
+
+## High-level description of the solution
+* Implement a new `AssemblyLoadContext` which will provide the isolation boundary and act as the "root" for the component. It can be enabled for unloading.  
+* The new load context is initialized by specifying the full path to the main assembly of the component to load.  
+* It will look for the `.deps.json` next to that assembly to determine its dependencies. Lack of `.deps.json` will be treated in the same way it is today for executable apps - that is all the assemblies in the same folder will be used as dependencies.
+* Parsing and understanding of the `.deps.json` will be performed by the same host components which do this for executable apps (so same behavior/quirks/bugs, very little code duplication).
+* If the component has `.runtimeconfig.json` it will only be used to verify runtime version and provide probing paths.
+* The load context will determine a list of assemblies similar to TPA and list of native search paths. These will be used to resolve any assembly binding events in the load context.
+* If the load context can't resolve an assembly bind event, it will fallback to the parent load context (the app)
+
+## Important implications and limitations
+* Only framework dependent components will be supported. Self-contained components will not be supported even if there was a way to produce them.
```

Comment:
Is there a technical reason for this limitation?  In general the host is able to resolve self contained references just as well as framework dependent.

Reply:
This restriction is a combination of
* Inability to override certain system libraries (like CoreLib)
* Untested behavior of overriding other system libraries.
* Implementing a simple solution to start with -- with the understanding that this solution will be sufficient for mainstream plugins.

In the appmodel meeting I think we said we should start with this restriction and can allow relaxations if there is a use-case.

Reply:
I'm just going to reword what @swaroop-sridhar  sais:
* Can can't override some assets - S.P.CoreLib, the runtime itself. So we would have to special case those.
* In fact overriding anything in `Microsoft.NETCore.App` is dangerous. Assemblies in `Microsoft.NETCore.App` can have direct references into S.P.CoreLib. And since S.P.CoreLib will be only one, allowing different versions of assemblies which depend on it may cause breaks. As far as I know we are free to introduce breaking changes to the S.P.CoreLib surface area at any time we want since only assemblies in `Microsoft.NETCore.App` can refer to it and we service that as a whole.
* The above would require us to somehow recognize anything from `Microsoft.NETCore.App`, which is actually rather tricky to do. Self-contained apps don't carry good description which assembly came from which framework. So we would porbably need a non-trivial set of hardcoded rules around this.
* For higher level frameworks this is technically possible and should not break anything really. I think it's counter intuitive in msot cases so we should make this "hard", but possible. Which it actually is. In theory I can create an app which depends only on one framework `Microsoft,NETCore.App` but carries a full copy of ASPNET framework as "self-contained". It would probably be pretty hard to produce this with current SDK, but one can craft a `.deps.json` like that if needed. - for this case I think we should wait for a specific user scenario before we go there. It might be a lot of work without a clear need right now.


All that said - it is definitely possible to include a one-of assembly from `Microsoft.NETCore.App` in the component itself and it will be used over the one from the framework. Just like it works today with apps.

There's also another issue with with SDK. For a NetStandard classlib the SDK actually ignores any attempts to make it self-contained and will always produce what is effectively a framework dependent thing. For NetCoreApp classlib, it is possible to produce a self-contained publish, but it's really weird (it gets `apphost.exe` - and I mean literally `apphost.exe`, it's not renamed, regardless of the name of the classlib). The apphost.exe is actually not even bound with the managed dll, so it shows a really bad error message.

Reply:
> For NetCoreApp classlib, it is possible to produce a self-contained publish, but it's really weird

Sounds like something worth filling an SDK issue for.

Reply:
Right. Filed:
dotnet/sdk#2550 and dotnet/sdk#2549.

Reply:
In a sense these must be 'app dependent components" where the app is required to provide the .NET runtime and related framework.